packages/store/src/Buffer.sol

```diff
@@ -0,0 +1,368 @@
+    uint256 _ptr = ptr(self);
+    assembly {
+      _length := mload(sub(_ptr, 32))
+    }
+  }
+
+  /**
+   * @dev Returns the capacity in bytes.
+   */
+  function capacity(Buffer self) internal pure returns (uint128 _capacity) {
+    return uint128(Buffer.unwrap(self));
+  }
+
+  /**
+   * @dev Read a 32 bytes from the buffer starting at the given offset (without checking for overflows)
+   */
+  function read32(Buffer self, uint256 offset) internal pure returns (bytes32 value) {
+    uint256 _ptr = ptr(self);
+    assembly {
+      value := mload(add(_ptr, offset))
+    }
+  }
+
+  /**
+   * @dev Read a 8 bytes from the buffer starting at the given offset (without checking for overflows)
+   */
+  function read8(Buffer self, uint256 offset) internal pure returns (bytes8 value) {
+    return bytes8(read32(self, offset));
+  }
+
+  /**
+   * @dev Read a 1 byte from the buffer starting at the given offset
+   */
+  function read1(Buffer self, uint256 offset) internal pure returns (bytes1 value) {
+    return bytes1(read32(self, offset));
+  }
+
+  // TODO: add more typed utils to read from the buffer
+
+  /**
+   * @dev Appends the given bytes to the buffer (checking for overflows)
+   */
+  function append(Buffer self, bytes memory data) internal pure {
+    uint128 _newLength = length(self) + uint128(data.length);
+    checkCapacity(self, _newLength);
+    appendUnchecked(self, data);
+  }
+
+  /**
+   * @dev Appends bytes32 to the buffer (checking for overflows)
+   */
+  function append(Buffer self, bytes32 data) internal pure {
+    append(self, data, 32);
+  }
+
+  /**
+   * @dev Appends _dataLength of the given bytes32 to the buffer (checking for overflows)
+   */
+  function append(
+    Buffer self,
+    bytes32 data,
+    uint128 _dataLength
+  ) internal pure {
+    uint128 _newLength = length(self) + _dataLength;
+    checkCapacity(self, _newLength);
+    appendUnchecked(self, data, _dataLength);
+  }
+
+  /**
+   * @dev Appends _dataLength of the given bytes32 to the buffer (without checking for overflows)
+   */
+  function appendUnchecked(
+    Buffer self,
+    bytes32 data,
+    uint128 _dataLength
+  ) internal pure {
+    uint256 _ptr = ptr(self);
+    uint128 _length = length(self);
+
+    // Update the current buffer length
+    _setLengthUnchecked(self, _length + _dataLength);
+
+    // Copy over given data to the buffer
+    assembly {
+      mstore(add(_ptr, _length), data)
+    }
+  }
+
+  /**
+   * @dev Appends the given bytes to the buffer (without checking for overflows)
+   */
+  function appendUnchecked(Buffer self, bytes memory data) internal pure {
+    uint256 _ptr = ptr(self);
+    uint128 _dataLength = uint128(data.length);
+    uint128 _length = length(self);
+    uint128 _newLength = _length + uint128(data.length);
+
+    // Update the current buffer length
+    _setLengthUnchecked(self, _newLength);
+
+    // Copy over given data to the buffer
+    assembly {
```

Comment:
You copy in 32 byte chunks, whereas the memory isn't word-aligned, so for if e.g. `_dataLength == 8`, memory corruption can happen.
[A verbose but pretty efficient alternative](https://gist.github.com/dk1a/6944af1557268e86b6d9d7ad801a93f9), for long bytes it can even costs less gas

Reply:
Good point! Mind changing this in a follow up PR?